// @sap/cds-dk/lib/init/constants.js
module.exports = {
    IS_WIN: require('os').platform() === 'win32',

    COMMAND_INIT: 'init',
    COMMAND_ADD: 'add',

    MAVEN_ARCHETYPE_VERSION: '2.8.1',

    OPTIONS: Object.freeze({
        JAVA: 'java',
        JAVA_MVN: 'java:mvn',
        JAVA_PACKAGE: 'java:package',
        NODEJS: 'nodejs',
        SAMPLES: 'samples', // deprecated
        TINY_SAMPLE: 'tiny-sample',
        SAMPLE: 'sample',
        PIPELINE: 'pipeline',
        HELM_UNIFIED_RUNTIME: 'helm-unified-runtime',
        HELM: 'helm',
        MTA: 'mta',
        CONNECTIVITY: 'connectivity',
        CF_MANIFEST: 'cf-manifest',
        MTX: 'mtx',
        MULTITENANCY: 'multitenancy',
        TOGGLES: 'toggles',
        EXTENSIBILITY: 'extensibility',
        XSUAA: 'xsuaa',
        HANA: 'hana',
        POSTGRES: 'postgres',
        SQLITE: 'sqlite',
        H2: 'h2',
        LIQUIBASE: 'liquibase',
        LOCAL_MESSAGING: 'local-messaging',
        FILE_BASED_MESSAGING: 'file-based-messaging',
        ENTERPRISE_MESSAGING: 'enterprise-messaging',
        ENTERPRISE_MESSAGING_SHARED: 'enterprise-messaging-shared',
        REDIS_MESSAGING: 'redis-messaging',
        KAFKA: 'kafka',
        APPROUTER: 'approuter',
        DESTINATION: 'destination',
        DATA: 'data',
        APPLICATION_LOGGING: 'application-logging',
        LINT: 'lint',
        LINT_GLOBAL: 'lint:global',
        LINT_DEV: 'lint:dev',
        TYPER: 'typer',
        HTML5_REPO: 'html5-repo',
        NOTIFICATIONS: 'notifications'
    }),

    REGEX_JAVA_PACKAGE: /^[a-z_][a-z0-9_]*(\.[a-z_][a-z0-9_]*)*$/g,
    // copied from mbt error message
    REGEX_JAVA_VERSION: /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\+[0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*)?$/,
    REGEX_PROJECT_NAME: /[^0-9a-zA-Z_-]+/g,

    URLS: Object.freeze({
        PIPELINE_HELP: 'https://sap.github.io/jenkins-library/',
        SAMPLES: 'https://github.com/SAP-samples/cloud-cap-samples',
        CAPIRE: 'https://cap.cloud.sap',
        MAVEN_ARCHETYPE_HELP: `https://cap.cloud.sap/docs/java/developing-applications/building#the-maven-archetype`,
        MAVEN_INSTALL_HELP: `https://maven.apache.org/install.html`
    }),

    PROJECT_FILES: ['package.json', 'pom.xml', '.cdsrc.json']
}
