// @sap/cds-dk/lib/init/template/redis-messaging/index.js
const cds = require('../../../cds')
const { read } = cds.utils
const { readProject } = require('../../projectReader')
const { merge } = require('../../merge')
const { srv4, redis } = require('../../registries/mta')

module.exports = class RedisMessagingTemplate extends require('../../plugin') {

  async canRun() {
    const { hasMta, hasHelm, hasHelmUnifiedRuntime, isJava } = readProject()
    const onlyHelm = !hasMta && (hasHelmUnifiedRuntime || hasHelm)
    if (isJava)   throw `'cds add redis-messaging' is not available for Java yet`
    if (onlyHelm) throw `'cds add redis-messaging' is not available for Kyma yet`
    return true
  }

  static hasInProduction(env) {
    return env.requires?.messaging?.kind === 'redis-messaging'
  }

  async run() {
    const project = readProject()
    const { configFile } = project
    project.shortcut = !(await read(configFile)).cds?.requires?.messaging?.kind
    await merge(__dirname, 'files/package.json.hbs').into(configFile, { project })
  }

  async combine() {
    const project = readProject()
    const { hasMta, srvPath } = project

    if (hasMta) {
      const srv = srv4(srvPath)
      await merge(__dirname, 'files/mta.yaml.hbs').into('mta.yaml', {
        project,
        additions: [srv, redis],
        relationships: [{
          insert: [redis, 'name'],
          into: [srv, 'requires', 'name']
        }]
      })
    }
  }
}
