// @sap/eslint-plugin-cds/lib/rules/auth-use-requires.js
module.exports = {
  meta: {
    schema: [{/* to avoid deprecation warning for ESLint 9 */}],
    docs: {
      description: 'Use `@requires` instead of `@restrict.to` in actions and services with unrestricted events.',
      category: 'Model Validation',
      recommended: true,
      version: '2.4.1'
    },
    hasSuggestions: true,
    messages: {
      InvalidItem: "Invalid item '{{invalid}}'. Did you mean '{{candidates}}'?",
      ReplaceItemWith: "Replace '{{invalid}}' with '{{candidates}}'"
    },
    type: 'problem',
    model: 'inferred'
  },
  create (context) {
    return {
      service: checkRestrict,
      action: checkRestrict
    }

    function checkRestrict (e) {
      if (e && e['@restrict'] && typeof e['@restrict'] === 'object') {
        for (const entry of e['@restrict']) {
          const keys = Object.keys(entry)
          if (keys.includes('to') && keys.includes('grant') && entry.grant === '*') {
            context.report({
              message: `Use \`@requires\` instead of \`@restrict.to\` at ${e.kind} \`${e.name}\`.`,
              node: context.getNode(e),
              file: e.$location.file
            })
          }
        }
      }
    }
  }
}
