// @sap/cds-dk/lib/build/constants.js
exports.OUTPUT_MODE = "outputMode"
exports.OUTPUT_MODE_NONE = 0x0
exports.OUTPUT_MODE_FILESYSTEM = 0x1
exports.OUTPUT_MODE_RESULT = 0x2

exports.DEFAULT_SRC_FOLDER = '.'
exports.BUILD_TASK_NPM_SCOPE = "@sap"
exports.BUILD_TASK_PREFIX = exports.BUILD_TASK_NPM_SCOPE + "/cds/build"
exports.BUILD_TASK_JAVA = "java"
exports.BUILD_TASK_JAVA_CF = "java-cf"
exports.BUILD_TASK_NODEJS = "nodejs"
exports.BUILD_TASK_NODE_CF = "node-cf"
exports.BUILD_TASK_HANA = "hana"
exports.BUILD_TASK_FIORI = "fiori"
exports.BUILD_TASK_MTX = "mtx"
exports.BUILD_TASK_MTX_SIDECAR = "mtx-sidecar"
exports.BUILD_TASK_MTX_EXTENSION = "mtx-extension"
exports.BUILD_TASKS = [this.BUILD_TASK_JAVA, this.BUILD_TASK_JAVA_CF, this.BUILD_TASK_NODEJS, this.BUILD_TASK_NODE_CF, this.BUILD_TASK_HANA, this.BUILD_TASK_FIORI, this.BUILD_TASK_MTX, this.BUILD_TASK_MTX_SIDECAR, this.BUILD_TASK_MTX_EXTENSION]

exports.ODATA_VERSION_V2 = "v2"
exports.ODATA_VERSION_V4 = "v4"

exports.CONTENT_LANGUAGE_BUNDLES = "contentLanguageBundles"     // create i18n.json language bundles
exports.CONTENT_DEFAULT_CSN = "contentDefaultCsn"               // create default CSN format flavor: "inferred"
exports.FLAVOR_LOCALIZED_EDMX = "flavorLocalizedEdmx"           // create localized EDMX languages based on existing properties files
exports.CONTENT_EDMX = "contentEdmx"                            // create EDMX for required languages - ENABLED by default
exports.CONTENT_LOCALIZED_EDMX = "contentLocalizedEdmx"         // create EDMX for required languages - ENABLED for Java by default
exports.CONTENT_PACKAGE_JSON = "contentPackageJson"             // create package.json file if not existing, or modify existing package.json - ENABLED by default
exports.CONTENT_PACKAGELOCK_JSON = "contentPackageLockJson"     // copy package-lock.json file into deployment folder - ENABLED by default
exports.CONTENT_HDBTABLEDATA = "contentHdbtabledata"            // create .hdbtabledata files for .csv files if not already existing - ENABLED by default
exports.CONTENT_NPMRC = "contentNpmrc"                          // copy .npmrc file into deployment folder - ENABLED by default
exports.CONTENT_CDSRC_JSON = "contentCdsrcJson"                 // copy .cdsrc.json file into deployment folder - ENABLED by default
exports.CONTENT_NODE_MODULES = "contentNodeModules"             // copy node_modules folder into deployment folder - DISABLED by default
exports.CONTENT_ENV = "contentEnv"                              // copy .env file into deployment folder - DISABLED by default
exports.CONTENT_DEFAULT_ENV_JSON = "contentDefaultEnvJson"      // copy default-env.json file into deployment folder - DISABLED by default
exports.CONTENT_WS = "ws"                                       // determine model paths including submodules + CONTENT_SUBMODULES_HANA - DISABLED by default

exports.CSV_FILE_TARGET = "csvFileTarget"                       // target folder when copying CSV files to the deployment target folder ./db/src/gen/*, default is 'data'
exports.CSV_FILE_DETECTION = "csvFileDetection"                 // detect CSV files based on CDS model location - ENABLED by default
exports.EXTENSION_POINT_VALIDATION = "extensionPointValidation" // validate extension point restrictions defined by the extended SaaS app - ENABLED by default
exports.MTX_SIDECAR_DB_VALIDATION = "mtxSidecarDbValidation"    // validate consistency of app and sidecar db configuration - ENABLED by default
exports.CONTINUE_UNRESOLVED_SCHEMA_CHANGES = "continueUnresolvedSchemaChanges" // build doesn't fail if unresolved HANA schema changes are detected
exports.IGNORE_DEFAULT_MODELS = "ignoreDefaultModels"           // do not add default models (@sap/cds-*, fts/*) to existing build task model options

exports.FOLDER_GEN = "gen"
exports.FILE_EXT_CDS = ".cds"
exports.MTX_SIDECAR_FOLDER = "mtx/sidecar"                      // default name of the mtx sidecar folder
exports.DEFAULT_CSN_FILE_NAME = "csn.json"
// REVISIT: the models are not required if a custom server.js file is used for MTX bootstrap
exports.NODEJS_MODEL_EXCLUDE_LIST = ['@sap/cds/srv/mtx', '@sap/cds-mtxs/srv/bootstrap']

exports.CDS_CONFIG_PATH_SEP = "/"
exports.SEVERITY_ERROR = "Error"
exports.SEVERITY_WARNING = "Warning"
exports.SEVERITY_INFO = "Info"
exports.SEVERITY_DEBUG = "Debug"
exports.SEVERITIES = [exports.SEVERITY_ERROR, exports.SEVERITY_WARNING, exports.SEVERITY_INFO, exports.SEVERITY_DEBUG]
exports.LOG_LEVEL_ERROR = "error"
exports.LOG_LEVEL_WARN = "warn"
exports.LOG_LEVEL_INFO = "info"
exports.LOG_LEVEL_DEBUG = "debug"
exports.LOG_LEVELS = [exports.LOG_LEVEL_ERROR, exports.LOG_LEVEL_WARN, exports.LOG_LEVEL_INFO, exports.LOG_LEVEL_DEBUG]
