// @sap/cds-dk/lib/init/template/lint/files/.eslint/rules/no-entity-moo.js
module.exports = require("@sap/eslint-plugin-cds").createRule({
  meta: {
    docs: {
      description: `Demonstrates how to implement a model validation check.

           __________________-_______
          < Entity 'Moo' not allowed! >
           --------------------------
                  \   ^__^
                   \  (oo)\_______
                      (__)\       )\/\
                          ||----w |
                          ||     ||`,
      version: "1.0.0"
    },
    fixable: "code"
  },
  create: function (context) {
    return { entity: checkForMoo };

    function checkForMoo(e) {
      const entityName = e.name.split(".").pop();
      if (entityName === "Moo") {
        const loc = context.getLocation(entityName, e);
        const fix = (fixer) => {
          const rangeEnd = context.getSourceCode().getIndexFromLoc({
            line: loc.end.line,
            column: loc.end.column
          });
          return fixer.insertTextAfterRange([0, rangeEnd], "n");
        };
        context.report({
          message: "Entity 'Moo' not allowed!",
          node: context.getNode(e),
          fix
        });
      }
    }
  }
})
